src/bitcoin/block.h

```diff
@@ -21,47 +27,137 @@ namespace bitcoin {
+    /** Height of the block; length of the chain from genesis */
+    int32_t nHeight;
+    /**
+     * Epochs are 5040 blocks long; each block points to the prev block of the
+     * first block within an epoch
+     */
+    uint256 hashEpochBlock;
+    /** Merkle root of the txs in the block */
+    uint256 hashMerkleRoot;
+    /** Hash of the extended metadata of the block */
+    uint256 hashExtendedMetadata;
 
     CBlockHeader() { SetNull(); }
 
     ADD_SERIALIZE_METHODS
 
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream &s, Operation ser_action) {
-        READWRITE(this->nVersion);
         READWRITE(hashPrevBlock);
-        READWRITE(hashMerkleRoot);
-        READWRITE(nTime);
         READWRITE(nBits);
+        READWRITE(vTime);
+        READWRITE(nReserved);
         READWRITE(nNonce);
+        READWRITE(nHeaderVersion);
+        READWRITE(vSize);
+        READWRITE(nHeight);
+        READWRITE(hashEpochBlock);
+        READWRITE(hashMerkleRoot);
+        READWRITE(hashExtendedMetadata);
     }
 
     void SetNull() {
-        nVersion = 0;
-        hashPrevBlock.SetNull();
-        hashMerkleRoot.SetNull();
-        nTime = 0;
+        hashPrevBlock = BlockHash();
         nBits = 0;
+        vTime.fill(0);
+        nReserved = 0;
         nNonce = 0;
+        nHeaderVersion = 0;
+        vSize.fill(0);
+        nHeight = 0;
+        hashEpochBlock.SetNull();
+        hashMerkleRoot.SetNull();
+        hashExtendedMetadata.SetNull();
     }
 
     bool IsNull() const { return (nBits == 0); }
 
-    uint256 GetHash() const;
+    BlockHash GetHash() const;
+
+    /**
+     * Block time as advertised in the block header.
+     * Might not correspond to actual mined or received time.
+     */
+    int64_t GetBlockTime() const {
+        return uint64_t(vTime[0]) | (uint64_t(vTime[1]) << 8) |
+               (uint64_t(vTime[2]) << 16) | (uint64_t(vTime[3]) << 24) |
+               (uint64_t(vTime[4]) << 32) | (uint64_t(vTime[5]) << 40);
+    }
+
+    /** Set the block time as advertised in the block header. */
+    void SetBlockTime(uint64_t nTime) {
+        vTime = {{
+            uint8_t((nTime & 0x0000000000ff)),
+            uint8_t((nTime & 0x00000000ff00) >> 8),
+            uint8_t((nTime & 0x000000ff0000) >> 16),
+            uint8_t((nTime & 0x0000ff000000) >> 24),
+            uint8_t((nTime & 0x00ff00000000) >> 32),
+            uint8_t((nTime & 0xff0000000000) >> 40),
+        }};
+    }
+
+    /**
+     * Block size, with full encoding, including header, metadata and txs.
+     * This doesn't measure anything, so could diverge from the actual size of
+     * an encoded CBlock.
+     */
+    uint64_t GetSize() const {
+        return uint64_t(vSize[0]) | (uint64_t(vSize[1]) << 8) |
+               (uint64_t(vSize[2]) << 16) | (uint64_t(vSize[3]) << 24) |
+               (uint64_t(vSize[4]) << 32) | (uint64_t(vSize[5]) << 40) |
+               (uint64_t(vSize[6]) << 48);
```

Comment:
Nit: Extra parens are unnecessary due to operator precedence.  I find this easier to read due to less paren-noise, but it's a matter of taste really...

```suggestion
        return uint64_t(vSize[0])       | uint64_t(vSize[1]) <<  8 |
               uint64_t(vSize[2]) << 16 | uint64_t(vSize[3]) << 24 |
               uint64_t(vSize[4]) << 32 | uint64_t(vSize[5]) << 40 |
               uint64_t(vSize[6]) << 48;
```